File: apps/launches/moarchy_launches/store.py
"""What survives the app being killed: the stars, and the last launches seen.

Two files, and the split is the whole design of this module.

`favourites.json` is a list of launch ids and is the only thing in this app a
person has made. It is written the instant a star is tapped, through a
temporary file and a rename, because on a phone "the process was killed a
moment after the tap" is the ordinary case: the compositor reclaims
backgrounded apps, and nothing asks first.

`upcoming.json` is the last answer Launch Library gave. It exists so that an
app opened on a train with no signal opens on countdowns rather than on an
apology, and it is disposable by definition -- every byte of it is replaced
by the next successful fetch. It is written when the window leaves the
screen rather than on each refresh, because a refresh around T-0 is every
two minutes and this file is a hundred kilobytes: an app left open through
a countdown would otherwise fsync several megabytes onto the phone's flash
to save a copy of something it already had in memory.

The launches are cached; *which ones you fly to* is not. Nothing here holds
a ticket, a seat, or a stream URL, and the app never asks for one -- the
whole file is public data plus a list of names, which is what makes leaving
it unencrypted in ~/.local/share an honest decision rather than an oversight.
"""

from __future__ import annotations

import json
import os
import time
from pathlib import Path

from .launches import Launch

SCHEMA = 1

FAVOURITES = "favourites.json"
UPCOMING = "upcoming.json"

# A guard on the file rather than on the person. Nobody stars two hundred
# launches; a favourites file that says they do has been written by something
# other than this app, and the list is what every refresh iterates.
MAX_FAVOURITES = 200


def data_dir() -> Path:
    """Where this app keeps its two files.

    Overridable, which is what makes the tests and the screenshot harness
    possible without touching anybody's real stars.
    """
    override = os.environ.get("MOARCHY_LAUNCHES_DIR")
    if override:
        return Path(override)
    base = os.environ.get("XDG_DATA_HOME") or str(Path.home() / ".local" / "share")
    return Path(base) / "moarchy-launches"


def _write(path: Path, payload: dict) -> None:
    """Temporary file, fsync, rename. See the header."""
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(".tmp")
    with tmp.open("w", encoding="utf-8") as fh:
        json.dump(payload, fh, ensure_ascii=False, indent=1)
        fh.flush()
        os.fsync(fh.fileno())
    os.replace(tmp, path)


def _read(path: Path) -> dict | None:
    """The file as a table, or None for anything else.

    A file that cannot be parsed is moved aside rather than left in place to
    be overwritten by the next save, which is the one operation that would
    destroy whatever the user actually had. The same thing Habits does, for
    the same reason: the broken copy is the only evidence of what went wrong.
    """
    try:
        raw = path.read_text(encoding="utf-8")
    except OSError:
        return None
    try:
        data = json.loads(raw)
    except ValueError:
        try:
            path.rename(path.with_suffix(f".broken-{int(time.time())}.json"))
        except OSError:
            pass
        return None
    return data if isinstance(data, dict) else None


class Store:
    """The stars, the cached upcoming list, and the two files they live in."""

    def __init__(self, directory: Path | str | None = None) -> None:
        self.dir = Path(directory) if directory else data_dir()
        # Launch ids, in the order they were starred. See `starred()`.
        self.favourites: list[str] = []
        self.launches: list[Launch] = []
        # Wall-clock seconds, because this one is compared against a clock the
        # user can see. Everything about *when to fetch again* is monotonic and
        # lives in the window.
        self.fetched = 0.0

    # --- files -----------------------------------------------------------

    @property
    def favourites_path(self) -> Path:
        return self.dir / FAVOURITES

    @property
    def upcoming_path(self) -> Path:
        return self.dir / UPCOMING

    def load(self) -> None:
        self._load_favourites()
        self._load_upcoming()

    def _load_favourites(self) -> None:
        data = _read(self.favourites_path)
        raw = data.get("favourites") if data else None
        if not isinstance(raw, list):
            self.favourites = []
            return
        seen: list[str] = []
        for entry in raw:
            if isinstance(entry, str) and entry and entry not in seen:
                seen.append(entry)
        self.favourites = seen[:MAX_FAVOURITES]

    def _load_upcoming(self) -> None:
        data = _read(self.upcoming_path)
        if not data:
            return
        records = data.get("launches")
        if not isinstance(records, list):
            return
        launches = []
        for record in records:
            try:
                launches.append(Launch.from_dict(record))
            except (ValueError, TypeError):
                continue
        self.launches = launches
        fetched = data.get("fetched")
        self.fetched = float(fetched) if isinstance(fetched, (int, float)) else 0.0

    def save_favourites(self) -> None:
        _write(self.favourites_path, {"schema": SCHEMA, "favourites": self.favourites})

    def save_upcoming(self) -> None:
        if not self.launches:
            # Nothing to cache, and writing an empty list would turn "we have
            # never fetched" into "the pad is empty" on the next launch.
            return
        _write(
            self.upcoming_path,
            {
                "schema": SCHEMA,
                "fetched": self.fetched,
                "launches": [launch.to_dict() for launch in self.launches],
            },
        )

    # --- the stars -------------------------------------------------------

    def is_favourite(self, launch_id: str) -> bool:
        return launch_id in self.favourites

    def toggle(self, launch_id: str) -> bool:
        """Star or unstar a launch. Returns whether it is now starred.

        Appends rather than inserts, which is what makes the starred page read
        in the order things were added to it.
        """
        if launch_id in self.favourites:
            self.favourites.remove(launch_id)
            return False
        if len(self.favourites) >= MAX_FAVOURITES:
            return False
        self.favourites.append(launch_id)
        return True

    def get(self, launch_id: str) -> Launch | None:
        for launch in self.launches:
            if launch.id == launch_id:
                return launch
        return None

    def starred(self) -> list[Launch]:
        """The starred launches, in the order they were starred.

        Not in NET order, which is the obvious alternative and is wrong on a
        phone: a watchlist sorted by countdown reorders itself under a thumb
        that is halfway down it, and the one ordering a person can actually
        control -- without a drag handle this app has no room for -- is when
        they added each one.

        A star with no launch behind it is left out rather than drawn as a
        row of dashes. That happens when a launch has flown off the upcoming
        list, and a row that says nothing about a launch is worse than the
        launch simply being gone. There is no second request for it: fifteen
        calls an hour cannot afford Coins' "ask by name" exception.
        """
        known = {launch.id: launch for launch in self.launches}
        return [known[i] for i in self.favourites if i in known]

    # --- the list --------------------------------------------------------

    def replace(self, launches: list[Launch], *, fetched: float) -> None:
        """Take a fetch, keeping at most one row per launch, in arrival order.

        Arrival order *is* NET order: that is how the endpoint sorts, and it
        is how the upcoming page is drawn.
        """
        seen: dict[str, Launch] = {}
        order: list[str] = []
        for launch in launches:
            if launch.id not in seen:
                order.append(launch.id)
            seen[launch.id] = launch
        self.launches = [seen[i] for i in order]
        self.fetched = fetched

    def age(self, now: float | None = None) -> float:
        """How old the cached launches are, in seconds, never negative."""
        if self.fetched <= 0:
            return float("inf")
        return max(0.0, (now if now is not None else time.time()) - self.fetched)

File: plugins/org.moarchy.launches/Store.js
// The two files the GTK app already writes, in the same shape.
//
// ~/.local/share/moarchy-launches/favourites.json is the only thing in this
// app a person made; upcoming.json is the last answer Launch Library gave and
// is disposable by definition. A star tapped here is a star over there.
.pragma library

var SCHEMA = 1
var MAX_FAVOURITES = 200

function pad2(n) { return n < 10 ? "0" + n : String(n) }

function iso(ms) {
  if (!ms) return null
  var d = new Date(ms)
  return d.getUTCFullYear() + "-" + pad2(d.getUTCMonth() + 1) + "-" + pad2(d.getUTCDate())
         + "T" + pad2(d.getUTCHours()) + ":" + pad2(d.getUTCMinutes()) + ":" + pad2(d.getUTCSeconds()) + "Z"
}

function stamp(value) {
  if (typeof value !== "string" || !value.trim()) return 0
  var ms = Date.parse(value.trim())
  return isNaN(ms) ? 0 : ms
}

function parseFavourites(text) {
  var out = []
  if (!text || !String(text).trim()) return out
  var data
  try {
    data = JSON.parse(text)
  } catch (e) {
    return out
  }
  var raw = data && data.favourites
  if (!raw || raw.length === undefined) return out
  for (var i = 0; i < raw.length && out.length < MAX_FAVOURITES; i++) {
    var id = raw[i]
    if (typeof id === "string" && id && out.indexOf(id) < 0) out.push(id)
  }
  return out
}

function serializeFavourites(favourites) {
  return JSON.stringify({ schema: SCHEMA, favourites: favourites || [] }, null, 1)
}

function launchFrom(data) {
  if (!data || typeof data !== "object") return null
  var net = stamp(data.net)
  if (typeof data.id !== "string" || !data.id || !net) return null
  return {
    id: data.id,
    name: String(data.name || data.id),
    vehicle: String(data.vehicle || ""),
    agency: String(data.agency || ""),
    status_id: Math.round(Number(data.status_id) || 0),
    status: String(data.status || ""),
    net: net,
    precision: String(data.precision || ""),
    window_start: stamp(data.window_start),
    window_end: stamp(data.window_end),
    pad: String(data.pad || ""),
    location: String(data.location || ""),
    orbit: String(data.orbit || ""),
    mission_type: String(data.mission_type || ""),
    probability: (typeof data.probability === "number") ? data.probability : null,
    weather: String(data.weather || ""),
    hold: String(data.hold || ""),
    description: String(data.description || "")
  }
}

function launchTo(item) {
  return {
    id: item.id,
    name: item.name,
    vehicle: item.vehicle,
    agency: item.agency,
    status_id: item.status_id,
    status: item.status,
    net: iso(item.net),
    precision: item.precision,
    window_start: iso(item.window_start),
    window_end: iso(item.window_end),
    pad: item.pad,
    location: item.location,
    orbit: item.orbit,
    mission_type: item.mission_type,
    probability: item.probability,
    weather: item.weather,
    hold: item.hold,
    description: item.description
  }
}

// Returns { launches: [...], fetched: <epoch seconds> }.
function parseUpcoming(text) {
  var empty = { launches: [], fetched: 0 }
  if (!text || !String(text).trim()) return empty
  var data
  try {
    data = JSON.parse(text)
  } catch (e) {
    return empty
  }
  if (!data || !data.launches || data.launches.length === undefined) return empty
  var out = []
  for (var i = 0; i < data.launches.length; i++) {
    var item = launchFrom(data.launches[i])
    if (item) out.push(item)
  }
  return {
    launches: out,
    fetched: (typeof data.fetched === "number") ? data.fetched : 0
  }
}

function serializeUpcoming(launches, fetched) {
  var out = []
  for (var i = 0; i < (launches || []).length; i++) out.push(launchTo(launches[i]))
  return JSON.stringify({ schema: SCHEMA, fetched: fetched || 0, launches: out }, null, 1)
}

// Keeps at most one row per launch, in arrival order — which is NET order,
// because that is how the endpoint sorts.
function dedupe(launches) {
  var seen = {}
  var out = []
  for (var i = 0; i < (launches || []).length; i++) {
    var item = launches[i]
    if (seen[item.id]) continue
    seen[item.id] = true
    out.push(item)
  }
  return out
}

function toggle(favourites, id) {
  var out = []
  var found = false
  for (var i = 0; i < (favourites || []).length; i++) {
    if (favourites[i] === id) { found = true; continue }
    out.push(favourites[i])
  }
  if (found) return { favourites: out, starred: false, full: false }
  if (out.length >= MAX_FAVOURITES) return { favourites: out, starred: false, full: true }
  out.push(id)
  return { favourites: out, starred: true, full: false }
}
